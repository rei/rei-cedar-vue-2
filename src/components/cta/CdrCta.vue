<template>
  <a
    :class="[modifierClass, ctaClass, fullWidthClass]"
    :target="target"
    :rel="computedRel"
    :href="href"
  >
    <!-- @slot innerHTML on the inside of the cta component -->
    <slot />
    <svg
      xmlns="http://www.w3.org/2000/svg"
      viewBox="0 0 24 24"
      role="presentation"
      :class="$style['cdr-cta__icon']"
    >
      <!-- eslint-disable-next-line -->
      <path d="M16 12a.997.997 0 0 0-.288-.702l-5.005-5.005a1 1 0 0 0-1.414 1.414L13.585 12 9.29 16.295a1 1 0 0 0 1.417 1.412l4.98-4.98A.997.997 0 0 0 16 12z" />
    </svg>
  </a>
</template>

<script>
import modifier from 'mixinsdir/modifier';

export default {
  name: 'CdrCta',
  mixins: [modifier],
  props: {
    /**
      * Change the color of the cdr-cta button match different themes.
      */
    ctaStyle: {
      type: String,
      default: 'dark',
      validator: value => (['brand', 'dark', 'light', 'sale'].indexOf(value) >= 0) || false,
    },
    /**
     * Sets width to be 100%.
    */
    fullWidth: {
      type: Boolean,
      default: false,
      validator: value => typeof value === 'boolean',
    },
    href: {
      type: String,
      default: '#',
    },
    /** @ignore */
    target: String,
    /** @ignore */
    rel: String,
  },
  computed: {
    baseClass() {
      return 'cdr-cta';
    },
    ctaClass() {
      return this.modifyClassName(this.baseClass, this.ctaStyle);
    },
    fullWidthClass() {
      return this.fullWidth && !this.iconOnly ?
        this.modifyClassName(this.baseClass, 'full-width') : null;
    },
    computedRel() {
      if (this.target === '_blank') {
        return this.rel || 'noopener noreferrer';
      }
      return this.rel;
    },
  },
};
</script>

<style module>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrCta.pcss';
</style>
