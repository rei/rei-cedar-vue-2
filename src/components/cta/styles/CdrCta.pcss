/* ==========================================================================
  # CdrCta
  
  All values should map to variables in
  vars/CdrCta.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrCta
========================================================================== */

.cdr-cta {
  @include maple-utility-40;

  background-color: $cta-background-color;
  box-shadow: $cta-box-shadow;
  border: none;
  border-radius: $cta-border-radius;
  color: $cta-color;
  cursor: pointer;
  display: inline-flex;
  outline: none;
  padding: $inset-1-and-a-half-x-squish;
  padding-right: calc($space-1-and-a-half-x - $space-half-x);
  text-align: left;
  text-transform: uppercase;
  text-decoration: none;

  & > .cdr-cta__icon {
    align-self: center;
    display: inline-block;
    fill: inherit;
    height: $icon-size;
    margin-left: $space-quarter-x;
    vertical-align: middle;
    flex: 0 0 auto;
    width: $icon-size;
  }

  &:hover,
  &:active,
  &:focus {
    text-decoration: none;
  }

  &:focus {
    outline: $default-outline;
    outline-offset: 0;
  }

  /* Full Width
     ========== */
  &--full-width {
    justify-content: center;
    width: 100%;
  }

  &--brand {
    background-color: $cta-brand-background-color;
    box-shadow: $cta-brand-box-shadow;
    color: $cta-brand-color;
    fill: $cta-brand-color;

    &:hover,
    &:focus,
    &:active {
      background-color: $cta-brand-hover-background-color;
      box-shadow: $cta-brand-box-shadow;
      color: $cta-brand-hover-color;
      fill: $cta-brand-hover-color;
    }
  }

  &--dark {
    background-color: $cta-dark-background-color;
    box-shadow: $cta-dark-box-shadow;
    color: $cta-dark-color;
    fill: $cta-dark-color;

    &:hover,
    &:focus,
    &:active {
      background-color: $cta-dark-hover-background-color;
      box-shadow: $cta-dark-box-shadow;
      color: $cta-dark-hover-color;

      & > .cdr-cta__icon {
        fill: $cta-dark-hover-color;
      }
    }
  }

  &--light {
    background-color: $cta-light-background-color;
    box-shadow: $cta-light-box-shadow;
    color: $cta-light-color;
    fill: $cta-light-color;

    &:hover,
    &:focus,
    &:active {
      background-color: $cta-light-hover-background-color;
      box-shadow: $cta-light-box-shadow;
      color: $cta-light-hover-color;
      fill: $cta-light-hover-color;
    }

    &.cdr-cta--elevated {
      box-shadow: $cta-light-box-shadow, $prominence-elevated;
      
      &:hover,
      &:focus,
      &:active, {
        box-shadow: $cta-light-box-shadow, $prominence-elevated;
      }
    }
  }

  &--sale {
    background-color: $cta-sale-background-color;
    box-shadow: $cta-sale-box-shadow;
    color: $cta-sale-color;
    fill: $cta-sale-color;

    &:hover,
    &:focus,
    &:active {
      background-color: $cta-sale-hover-background-color;
      box-shadow: $cta-sale-box-shadow;
      color: $cta-sale-hover-color;
      fill: $cta-sale-hover-color;
    }
  }
  
  &--elevated {
    box-shadow: $prominence-elevated;

    &:hover,
    &:focus,
    &:active {
      box-shadow: $prominence-elevated;
    }
  }
}
