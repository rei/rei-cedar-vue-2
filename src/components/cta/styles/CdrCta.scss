/* ==========================================================================
  # CdrCta

  All values should map to variables in
  vars/CdrCta.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrCta
========================================================================== */

.cdr-cta {
  @include cdr-cta-base-mixin;

  & > .cdr-cta__icon {
    @include cdr-cta-base-icon-mixin;
  }

  /* Full Width
     ========== */
  &--full-width,
  &--fullWidth\@all {
    @include cdr-cta-full-width-mixin;
  }

  &--brand {
    @include cdr-cta-brand-mixin;
  }

  &--dark {
    @include cdr-cta-dark-mixin;
  }

  &--light {
    @include cdr-cta-light-mixin;
    &.cdr-cta--elevated {
      @include cdr-cta-light-elevated-mixin;
    }
  }

  &--sale {
    @include cdr-cta-sale-mixin;
  }

  &--elevated {
    @include cdr-cta-elevated-mixin;
  }

  /* Full Width @ breakpoint
    ========== */
  @include xs-mq-only {
    &--fullWidth\@xs {
      @include cdr-cta-full-width-mixin;
    }
  }

  @include sm-mq-only {
    &--fullWidth\@sm {
      @include cdr-cta-full-width-mixin;
    }
  }

  @include md-mq-only {
    &--fullWidth\@md {
      @include cdr-cta-full-width-mixin;
    }
  }

  @include lg-mq-only {
    &--fullWidth\@lg {
      @include cdr-cta-full-width-mixin;
    }
  }
}
