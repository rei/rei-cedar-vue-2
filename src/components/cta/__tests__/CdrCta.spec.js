import { shallowMount, mount } from '@vue/test-utils';
import CdrCta from 'componentdir/cta/CdrCta';

describe('CdrCta', () => {
  test('renders correctly', () => {
    const wrapper = mount(CdrCta);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('adds classes from ctaStyle prop', () => {
    const wrapper = shallowMount(CdrCta, {
      propsData: {
        ctaStyle: 'sale',
      },
    });

    // it's important that it add cta as a separate class because BEM!
    expect(wrapper.classes()).toContain('cdr-cta');
    expect(wrapper.classes()).toContain('cdr-cta--sale');
  });

  describe('full width class', () => {
    it('adds full width class for boolean', () => {
      const wrapper = shallowMount(CdrCta, {
        propsData: {
          fullWidth: true,
        },
      });

      expect(wrapper.classes()).toContain('cdr-cta--full-width');
    });

    it('adds full width class for breakpoint', () => {
      const wrapper = shallowMount(CdrCta, {
        propsData: {
          fullWidth: "@xs",
        },
      });

      expect(wrapper.classes()).toContain('cdr-cta--fullWidth@xs');
    });
  });

  it('sets target attr correctly', () => {
    const wrapper = shallowMount(CdrCta, {
      propsData: {
        target: '_self',
      },
    });
    expect(wrapper.attributes().target).toBe('_self');
  });

  it('sets a default href', () => {
    const wrapper = shallowMount(CdrCta);
    expect(wrapper.attributes().href).toBe('#');
  });

  it('sets rel attr correctly', () => {
    const wrapper = shallowMount(CdrCta, {
      propsData: {
        rel: 'nofollow',
      },
    });
    expect(wrapper.attributes().rel).toBe('nofollow');
  });

  it('computes target="_blank" rel attr correctly', () => {
    const wrapper = shallowMount(CdrCta, {
      propsData: {
        target: '_blank',
      },
    });
    expect(wrapper.attributes().rel).toBe('noopener noreferrer');
  });
});
