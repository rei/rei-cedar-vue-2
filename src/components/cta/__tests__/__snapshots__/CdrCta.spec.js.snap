// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrCta renders correctly 1`] = `
<a
  class="cdr-cta cdr-cta--dark"
  href="#"
>
  <svg
    class="cdr-cta__icon"
    role="presentation"
    viewBox="0 0 24 24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M16 12a.997.997 0 0 0-.288-.702l-5.005-5.005a1 1 0 0 0-1.414 1.414L13.585 12 9.29 16.295a1 1 0 0 0 1.417 1.412l4.98-4.98A.997.997 0 0 0 16 12z"
    />
  </svg>
</a>
`;
